<map>: Using `operator[]` on a `std::pmr::map` fails to compile when the mapped type is a `std::pair`.

**Describe the bug**
In the example below `tags[0];` triggers the compilation error. It compiles fine with Clang and GCC. It also compiles as C++17, if the mapped type is replaced with a `struct pair { int first, second; }` and notably when the mapped type is a `tuple` instead of a `pair`.

Maybe the problem is caused by `try_emplace()`: [Compiler explorer example](https://godbolt.org/z/joe78W4r9) (thank you @PJBoy)

**Command-line test case**
```
C:\Temp>type repro.cpp
#include <map>

void foo()
{
    std::pmr::map<int, std::pair<int, int>> tags;
    tags[0];
}

C:\Temp>cl /std:c++latest /c /EHsc /W4 /WX .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30037 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

/std:c++latest is provided as a preview of language features from the latest C++
working draft, and we're eager to hear about bugs and suggestions for improvements.
However, note that these features are provided as-is without support, and subject
to changes or removal as the working draft evolves. See
https://go.microsoft.com/fwlink/?linkid=2045807 for details.

repro.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2051): error C2672: 'uses_allocator_construction_args': no matching overloaded function found
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\type_traits(1584): note: see reference to function template instantiation 'auto std::uses_allocator_construction_args::<lambda_2>::operator ()<>(void) const' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\tuple(1007): note: see reference to function template instantiation 'decltype(auto) std::_Apply_impl<_Ty,std::tuple<>,>(_Callable &&,_Tuple &&,std::integer_sequence<size_t>)' being compiled
        with
        [
            _Ty=std::uses_allocator_construction_args::<lambda_2>,
            _Callable=std::uses_allocator_construction_args::<lambda_2>,
            _Tuple=std::tuple<>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2054): note: see reference to function template instantiation 'decltype(auto) std::apply<std::uses_allocator_construction_args::<lambda_2>,_Ty>(_Callable &&,_Tuple &&)' being compiled
        with
        [
            _Ty=std::tuple<>,
            _Callable=std::uses_allocator_construction_args::<lambda_2>,
            _Tuple=std::tuple<>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2102): note: see reference to function template instantiation 'auto std::uses_allocator_construction_args<_Ty,_Alloc,std::tuple<int &&>,std::tuple<>,0>(const _Alloc &,std::piecewise_construct_t,_Tuple1 &&,_Tuple2 &&) noexcept' being compiled
        with
        [
            _Ty=std::pair<const int,std::pair<int,int>>,
            _Alloc=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>,
            _Tuple1=std::tuple<int &&>,
            _Tuple2=std::tuple<>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xpolymorphic_allocator.h(274): note: see reference to function template instantiation '_Ty *std::uninitialized_construct_using_allocator<_Uty,std::pmr::polymorphic_allocator<_Newfirst>,const std::piecewise_construct_t&,std::tuple<int &&>,std::tuple<>>(_Ty *,const _Alloc &,const std::piecewise_construct_t &,std::tuple<int &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Ty=std::pair<const int,std::pair<int,int>>,
            _Uty=std::pair<const int,std::pair<int,int>>,
            _Newfirst=std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>,
            _Alloc=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(597): note: see reference to function template instantiation 'void std::pmr::polymorphic_allocator<_Newfirst>::construct<_Ty,const std::piecewise_construct_t&,std::tuple<int &&>,std::tuple<>>(_Uty *const ,const std::piecewise_construct_t &,std::tuple<int &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Newfirst=std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>,
            _Ty=std::pair<const int,std::pair<int,int>>,
            _Uty=std::pair<const int,std::pair<int,int>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(606): note: see reference to function template instantiation 'void std::pmr::polymorphic_allocator<_Newfirst>::construct<_Ty,const std::piecewise_construct_t&,std::tuple<int &&>,std::tuple<>>(_Uty *const ,const std::piecewise_construct_t &,std::tuple<int &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Newfirst=std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>,
            _Ty=std::pair<const int,std::pair<int,int>>,
            _Uty=std::pair<const int,std::pair<int,int>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xtree(811): note: see reference to function template instantiation 'void std::_Normal_allocator_traits<_Alloc>::construct<_Ty,const std::piecewise_construct_t&,std::tuple<int &&>,std::tuple<>>(_Alloc &,_Ty *,const std::piecewise_construct_t &,std::tuple<int &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Alloc=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>,
            _Ty=std::pair<const int,std::pair<int,int>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xtree(812): note: see reference to function template instantiation 'void std::_Normal_allocator_traits<_Alloc>::construct<_Ty,const std::piecewise_construct_t&,std::tuple<int &&>,std::tuple<>>(_Alloc &,_Ty *,const std::piecewise_construct_t &,std::tuple<int &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Alloc=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>,
            _Ty=std::pair<const int,std::pair<int,int>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\map(175): note: see reference to function template instantiation 'std::_Tree_temp_node<std::pmr::polymorphic_allocator<_Newfirst>>::_Tree_temp_node<const std::piecewise_construct_t&,std::tuple<_Ty &&>,std::tuple<>>(_Alnode &,std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *> *,const std::piecewise_construct_t &,std::tuple<_Ty &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Newfirst=std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>,
            _Ty=int,
            _Alnode=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\map(184): note: see reference to function template instantiation 'std::_Tree_temp_node<std::pmr::polymorphic_allocator<_Newfirst>>::_Tree_temp_node<const std::piecewise_construct_t&,std::tuple<_Ty &&>,std::tuple<>>(_Alnode &,std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *> *,const std::piecewise_construct_t &,std::tuple<_Ty &&> &&,std::tuple<> &&)' being compiled
        with
        [
            _Newfirst=std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>,
            _Ty=int,
            _Alnode=std::pmr::polymorphic_allocator<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *>>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\map(154): note: see reference to function template instantiation 'std::pair<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *> *,bool> std::map<int,std::pair<int,int>,std::less<int>,std::pmr::polymorphic_allocator<std::pair<const int,std::pair<int,int>>>>::_Try_emplace<int,>(_Keyty &&)' being compiled
        with
        [
            _Keyty=int
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\map(154): note: see reference to function template instantiation 'std::pair<std::_Tree_node<std::pair<const int,std::pair<int,int>>,void *> *,bool> std::map<int,std::pair<int,int>,std::less<int>,std::pmr::polymorphic_allocator<std::pair<const int,std::pair<int,int>>>>::_Try_emplace<int,>(_Keyty &&)' being compiled
        with
        [
            _Keyty=int
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\map(153): note: while compiling class template member function 'std::pair<int,int> &std::map<int,std::pair<int,int>,std::less<int>,std::pmr::polymorphic_allocator<std::pair<const int,std::pair<int,int>>>>::operator [](int &&)'
.\repro.cpp(7): note: see reference to function template instantiation 'std::pair<int,int> &std::map<int,std::pair<int,int>,std::less<int>,std::pmr::polymorphic_allocator<std::pair<const int,std::pair<int,int>>>>::operator [](int &&)' being compiled
.\repro.cpp(6): note: see reference to class template instantiation 'std::map<int,std::pair<int,int>,std::less<int>,std::pmr::polymorphic_allocator<std::pair<const int,std::pair<int,int>>>>' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2051): error C2783: 'auto std::uses_allocator_construction_args(const _Alloc &,_Types &&...) noexcept': could not deduce template argument for '__formal'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2021): note: see declaration of 'std::uses_allocator_construction_args'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2042): error C2672: 'std::make_tuple': no matching overloaded function found
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2054): error C2893: Failed to specialize function template 'std::tuple<_Unrefwrap_helper<decay<_Types>::type>::type...> std::make_tuple(_Types &&...)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\tuple(888): note: see declaration of 'std::make_tuple'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2054): note: With the following template arguments:
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2054): note: '_Types={const std::piecewise_construct_t &, std::tuple<_Ty &&>, void}'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2103): error C2672: 'std::apply': no matching overloaded function found
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2102): error C2893: Failed to specialize function template 'decltype(auto) std::apply(_Callable &&,_Tuple &&)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\tuple(1005): note: see declaration of 'std::apply'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2102): note: With the following template arguments:
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2102): note: '_Callable=std::uninitialized_construct_using_allocator::<lambda_1>'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30037\include\xmemory(2102): note: '_Tuple=void'
```

**Expected behavior**
No compilation error. 

**STL version**
```
Microsoft Visual Studio Community 2019
Version 16.10.0
```

--- a/stl/inc/xmemory
+++ b/stl/inc/xmemory
@@ -2096,6 +2096,18 @@ _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, _T
     }
 }
 
+template <class _Ty, class _Alloc, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+_NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al) noexcept;
+
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+_NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, _Uty1&& _Val1, _Uty2&& _Val2) noexcept;
+
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+_NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, const pair<_Uty1, _Uty2>& _Pair) noexcept;
+
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+_NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, pair<_Uty1, _Uty2>&& _Pair) noexcept;
+
 template <class _Ty, class _Alloc, class _Tuple1, class _Tuple2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
 _NODISCARD constexpr auto uses_allocator_construction_args(
     const _Alloc& _Al, piecewise_construct_t, _Tuple1&& _Tup1, _Tuple2&& _Tup2) noexcept {
@@ -2114,7 +2126,7 @@ _NODISCARD constexpr auto uses_allocator_construction_args(
             _STD forward<_Tuple2>(_Tup2)));
 }
 
-template <class _Ty, class _Alloc, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+template <class _Ty, class _Alloc, enable_if_t<_Is_specialization_v<_Ty, pair>, int>>
 _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al) noexcept {
     // equivalent to
     // return _STD uses_allocator_construction_args<_Ty>(_Al, piecewise_construct, tuple<>{}, tuple<>{});
@@ -2122,7 +2134,7 @@ _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al) no
         _STD uses_allocator_construction_args<typename _Ty::second_type>(_Al));
 }
 
-template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int>>
 _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, _Uty1&& _Val1, _Uty2&& _Val2) noexcept {
     // equivalent to
     // return _STD uses_allocator_construction_args<_Ty>(_Al, piecewise_construct,
@@ -2132,7 +2144,7 @@ _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, _U
         _STD uses_allocator_construction_args<typename _Ty::second_type>(_Al, _STD forward<_Uty2>(_Val2)));
 }
 
-template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int>>
 _NODISCARD constexpr auto uses_allocator_construction_args(
     const _Alloc& _Al, const pair<_Uty1, _Uty2>& _Pair) noexcept {
     // equivalent to
@@ -2143,7 +2155,7 @@ _NODISCARD constexpr auto uses_allocator_construction_args(
         _STD uses_allocator_construction_args<typename _Ty::second_type>(_Al, _Pair.second));
 }
 
-template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int> = 0>
+template <class _Ty, class _Alloc, class _Uty1, class _Uty2, enable_if_t<_Is_specialization_v<_Ty, pair>, int>>
 _NODISCARD constexpr auto uses_allocator_construction_args(const _Alloc& _Al, pair<_Uty1, _Uty2>&& _Pair) noexcept {
     // equivalent to
     // return _STD uses_allocator_construction_args<_Ty>(_Al, piecewise_construct,

--- a/tests/std/tests/P0475R1_P0591R4_uses_allocator_construction/test.cpp
+++ b/tests/std/tests/P0475R1_P0591R4_uses_allocator_construction/test.cpp
@@ -174,7 +174,7 @@ constexpr bool test_P0591R4() {
     return true;
 }
 
-void test_GH_2021() { // compile-only
+void test_GH_2021() { // COMPILE-ONLY
     pmr::map<int, pair<int, int>> tags;
     tags[0];
 }

--- a/tests/std/tests/P0475R1_P0591R4_uses_allocator_construction/test.cpp
+++ b/tests/std/tests/P0475R1_P0591R4_uses_allocator_construction/test.cpp
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include <cassert>
+#include <map>
 #include <memory>
 #include <scoped_allocator>
 #include <tuple>
@@ -171,6 +172,11 @@ constexpr bool test_P0591R4() {
     }
 
     return true;
+}
+
+void test_GH_2021() { // compile-only
+    pmr::map<int, pair<int, int>> tags;
+    tags[0];
 }
 
 int main() {

